protocols/gossipsub/src/behaviour.rs

```diff
@@ -516,6 +532,10 @@ where
             // possible to have a message that exceeds the RPC limit and is not caught here. A
             // warning log will be emitted in this case.
             return Err(PublishError::MessageTooLarge);
+
+            // mxinden: In case the message is still too large, how about emitting an event
```

Comment:
> As a GossipsubEvent or simply adding a log?

Yes, passing it back as a `GossipsubEvent`. Again, I haven't found the time to fully think this through, but from a first grasp this seems intuitive.

Reply:
I guess this comes down to how we expect the application to handle this. I would have thought it more intuitive that the `gossipsub.subscribe()` function itself returns a Result and the application knows immediately there is an error and handles it then, rather than handling errors during the poll. 

This saves the application from having some intermediate state between sending a subscription and knowing whether it was successful or not. I guess if we throw a `GossipsubEvent` on failure, we might want to throw the success case also?

Reply:
I am a bit confused why you are referring to `subscribe()` and `unsubscribe()` here. How does this comment on `publish()` relate to the subscription or unsubscription paths?

> `// It is possible to have a message that exceeds the RPC limit and is not caught here. A warning log will be emitted in this case.`

Do I understand correctly that this comment refers to the `poll` function in `handler.rs`, more concretely the handling of `GossipsubHandlerError::MaxTransmissionSize` when sending a message?

```rust
                Some(OutboundSubstreamState::PendingSend(mut substream, message)) => {
                    match Sink::poll_ready(Pin::new(&mut substream), cx) {
                        Poll::Ready(Ok(())) => {
                            match Sink::start_send(Pin::new(&mut substream), message) {
                                Ok(()) => {
                                    self.outbound_substream =
                                        Some(OutboundSubstreamState::PendingFlush(substream))
                                }
                                Err(GossipsubHandlerError::MaxTransmissionSize) => {
                                    error!("Message exceeded the maximum transmission size and was not sent.");
                                    self.outbound_substream =
                                        Some(OutboundSubstreamState::WaitingOutput(substream));
                                }
                                Err(e) => {
                                    error!("Error sending message: {}", e);
                                    return Poll::Ready(ProtocolsHandlerEvent::Close(e));
                                }
                            }
                        }
                        Poll::Ready(Err(e)) => {
                            error!("Outbound substream error while sending output: {:?}", e);
                            return Poll::Ready(ProtocolsHandlerEvent::Close(e));
                        }
                        Poll::Pending => {
                            self.outbound_substream =
                                Some(OutboundSubstreamState::PendingSend(substream, message));
                            break;
                        }
                    }
                }
```

My initial idea was to encapsulate the original oversized message in the `GossipsubHandlerError::MaxTransmissionSize` error, bubbling it up to the user. Thinking some more about it, this is likely confusing as (a) sending a oversized message to multiple peers would result in multiple such events for a single `publish` call and (b) as you elaborated in the comment on `publish` the limit can both be reached through messages and topics. With that in mind I would leave it as is with the error log as well as the early length check in `publish`.

> `// It is possible to have a message that exceeds the RPC limit and is not caught here.`

How can the message to be send still in itself exceed the `max_transmit_size` even though, for the purpose of checking, it is being encoded as a protobuf message here to check the length? Or are you referring to something different here with `message`?

Reply:
@AgeManning pinging you here, not because it is urgent, but due to the comment being marked as _outdated_, and thus maybe getting lost.